### PR TITLE
dpp: init at 10.0.29

### PR DESCRIPTION
**Pull request by xbz-24**
## Description of changes

I have successfully built and tested the dpp package on my x86_64-linux machine. I have tested the package on other platforms like Darwin, but that one needs further testing, I mostly developed using dpp library on x86_64 so I am sure at least this architecture is supported, any feedback or testing on platforms like darwin or other would be greatly appreciated 

links
https://github.com/brainboxdotcc/DPP
https://dpp.dev/

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/libraries/dpp/default.nix [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+, zlib
+, libsodium
+, gcc
+, pkg-config
+, autoPatchelfHook
+, lib
+}:
+let
+   libExt = if stdenv.hostPlatform.isDarwin then ".dylib" else ".so";
+in
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/dpp/default.nix:15]:
```suggestion

```
we already have that in stdenv.hostPlatform.extensions.sharedLibrary and I also moved it closer to the usage

File: pkgs/development/libraries/dpp/default.nix [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    pkg-config
+  ];
+
+  PKG_CONFIG_PATH = [
+    "${openssl.dev}/lib/pkgconfig"
+    "${libsodium.dev}/lib/pkgconfig"
+    "${libopus.dev}/lib/pkgconfig"
+  ];
+
+  cmakeFlags = [
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/dpp/default.nix:47]:
```suggestion
  cmakeFlags = let
    libExt = stdenv.hostPlatform.extensions.sharedLibrary;
  in [
```

File: pkgs/development/libraries/dpp/default.nix [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    libopus
+  ] ++ lib.optional (stdenv.hostPlatform.isLinux) autoPatchelfHook;
+
+  buildInputs = [
+    openssl
+    zlib
+    libsodium
+    libopus
+    gcc
+    pkg-config
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/dpp/default.nix:38]: pkg-config must go into nativeBuildInputs

File: pkgs/development/libraries/dpp/default.nix [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    cmake
+    libopus
+  ] ++ lib.optional (stdenv.hostPlatform.isLinux) autoPatchelfHook;
+
+  buildInputs = [
+    openssl
+    zlib
+    libsodium
+    libopus
+    gcc
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/dpp/default.nix:37]: gcc is already contained in the stdenv

File: pkgs/development/libraries/dpp/default.nix [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  src = fetchFromGitHub {
+    owner = "brainboxdotcc";
+    repo = "DPP";
+    rev = "v${version}";
+    sha256 = "BJMg3MLSfb9x/2lPHITeI3SWwW1OZVUUMVltTWUcw9I=";
+  };
+
+  nativeBuildInputs =  [
+    cmake
+    libopus
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/dpp/default.nix:29]: libopus should probably go to buildInputs as it is a library

File: pkgs/development/libraries/dpp/default.nix [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+in
+stdenv.mkDerivation rec {
+  pname = "dpp";
+  version = "10.0.29";
+
+  src = fetchFromGitHub {
+    owner = "brainboxdotcc";
+    repo = "DPP";
+    rev = "v${version}";
+    sha256 = "BJMg3MLSfb9x/2lPHITeI3SWwW1OZVUUMVltTWUcw9I=";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/dpp/default.nix:24]:
```suggestion
    hash = "sha256-BJMg3MLSfb9x/2lPHITeI3SWwW1OZVUUMVltTWUcw9I=";
```
please don't strip sri hashes

File: pkgs/development/libraries/dpp/default.nix [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    libopus
+    gcc
+    pkg-config
+  ];
+
+  PKG_CONFIG_PATH = [
+    "${openssl.dev}/lib/pkgconfig"
+    "${libsodium.dev}/lib/pkgconfig"
+    "${libopus.dev}/lib/pkgconfig"
+  ];
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/dpp/default.nix:45]: This shouldn't be required when pkg-config is in nativeBuildInputs

File: pkgs/development/libraries/dpp/default.nix [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+  cmakeFlags = [
+    "-DLibsodium_INCLUDE_DIR=${libsodium.dev}/include"
+    "-DLibsodium_LIBRARY=${libsodium.out}/lib"
+    "-DOPENSSL_ROOT_DIR=${openssl.dev}"
+    "-DOPENSSL_LIBRARIES=${openssl.out}/lib"
+    "-DOPENSSL_CRYPTO_LIBRARY=${openssl.out}/lib/libcrypto${libExt}"
+    "-DOPENSSL_SSL_LIBRARY=${openssl.out}/lib/libssl${libExt}"
+    "-DOPENSSL_INCLUDE_DIR=${openssl.dev}/include"
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/dpp/default.nix:54]: Isn't this automatically found by the build system? normally CMake is quite good doing this.

**Review by SuperSandro2000 — COMMENTED** [review]
Please squash the review and fixup commits into ``dpp: init at 10.0.29``

File: pkgs/development/libraries/dpp/default.nix [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  ];
+
+  meta = {
+    description = "D++ Discord C++ Library";
+    homepage = "https://github.com/brainboxdotcc/DPP";
+    license = lib.licenses.asl20;
+    maintainers = with lib.maintainers; [ renato ];
```
Comment (dotlambda) [inline pkgs/development/libraries/dpp/default.nix:61]: This should be you.

File: pkgs/top-level/all-packages.nix [hunk 10]
```diff
@@ -624,6 +624,8 @@ with pkgs;
 
   domine = callPackage ../tools/misc/domine { };
 
+  dpp = callPackage ../development/libraries/dpp { };
```
Comment (dotlambda) [inline pkgs/top-level/all-packages.nix:627]: Please use [`pkgs/by-name`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md).

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/dpp/default.nix [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    "-DLibsodium_LIBRARY=${libsodium.out}/lib"
+    "-DOPENSSL_ROOT_DIR=${openssl.dev}"
+    "-DOPENSSL_LIBRARIES=${openssl.out}/lib"
+    "-DOPENSSL_CRYPTO_LIBRARY=${openssl.out}/lib/libcrypto${libExt}"
+    "-DOPENSSL_SSL_LIBRARY=${openssl.out}/lib/libssl${libExt}"
+    "-DOPENSSL_INCLUDE_DIR=${openssl.dev}/include"
+  ];
+
+  meta = {
+    description = "D++ Discord C++ Library";
```
[the remark this answers is not in the document]
Comment (Xbz-24) [reply pkgs/development/libraries/dpp/default.nix]: I made it one sentence now

File: pkgs/development/libraries/dpp/default.nix (hunk 8 above)
Comment (Xbz-24) [reply pkgs/development/libraries/dpp/default.nix]: I removed every one of this flags except the ones with libExt

File: pkgs/development/libraries/dpp/default.nix (hunk 7 above)
Comment (Xbz-24) [reply pkgs/development/libraries/dpp/default.nix]: removed

File: pkgs/development/libraries/dpp/default.nix (hunk 3 above)
Comment (Xbz-24) [reply pkgs/development/libraries/dpp/default.nix:35]: done

File: pkgs/development/libraries/dpp/default.nix (hunk 1 above)
Comment (Xbz-24) [reply pkgs/development/libraries/dpp/default.nix]: thx, I moved it closer to the usage as requested

File: pkgs/development/libraries/dpp/default.nix (hunk 4 above)
Comment (Xbz-24) [reply pkgs/development/libraries/dpp/default.nix]: you are right, sorry

File: pkgs/development/libraries/dpp/default.nix (hunk 5 above)
Comment (Xbz-24) [reply pkgs/development/libraries/dpp/default.nix:26]: done

File: pkgs/development/libraries/dpp/default.nix (hunk 9 above)
Comment (Xbz-24) [reply pkgs/development/libraries/dpp/default.nix]: I changed it to  rgchavezchicoma is that alright

File: pkgs/top-level/all-packages.nix (hunk 10 above)
Comment (Xbz-24) [reply pkgs/top-level/all-packages.nix:627]: I tried doing this while following the structure from the readme
    dpp = callPackage ../by-name/dp/dpp/package.nix { };
  
  
    "pkgs/by-name/dp/dpp/package.nix" being the path where the modified derivation is
  
   but I did something wrong, can you help me identify my error?

Comment (dotlambda) [reply pkgs/top-level/all-packages.nix:627]: Completely remove the line from `all-packages.nix`.

File: pkgs/development/libraries/dpp/default.nix
Comment (dotlambda) [inline pkgs/development/libraries/dpp/default.nix:1]: Remove this file.

File: pkgs/by-name/dp/dpp/package.nix [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+   in [
+    "-DOPENSSL_CRYPTO_LIBRARY=${openssl.out}/lib/libcrypto${libExt}"
+    "-DOPENSSL_SSL_LIBRARY=${openssl.out}/lib/libssl${libExt}"
+  ];
+
+  meta = {
+    description = "Discord C++ Library";
+    homepage = "https://github.com/brainboxdotcc/DPP";
+    license = lib.licenses.asl20;
+    maintainers = with lib.maintainers; [ rgchavezchicoma ];
```
Comment (dotlambda) [inline pkgs/by-name/dp/dpp/package.nix:49]: You have to add yourself to https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix in a separate commit.

File: pkgs/by-name/dp/dpp/package.nix [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    libsodium
+    libopus
+    pkg-config
+  ];
+
+  cmakeFlags = let
+      libExt = stdenv.hostPlatform.extensions.sharedLibrary;
+   in [
+    "-DOPENSSL_CRYPTO_LIBRARY=${openssl.out}/lib/libcrypto${libExt}"
+    "-DOPENSSL_SSL_LIBRARY=${openssl.out}/lib/libssl${libExt}"
```
Comment (dotlambda) [inline pkgs/by-name/dp/dpp/package.nix:42]: Why are these needed?

File: pkgs/by-name/dp/dpp/package.nix [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+    libopus
+    pkg-config
+  ] ++ lib.optional (stdenv.hostPlatform.isLinux) autoPatchelfHook;
+
+  buildInputs = [
+    openssl
+    zlib
+    libsodium
+    libopus
+    pkg-config
```
Comment (dotlambda) [inline pkgs/by-name/dp/dpp/package.nix:35]:
```suggestion
```

File: pkgs/by-name/dp/dpp/package.nix (hunk 13 above)
Comment (Xbz-24) [reply pkgs/by-name/dp/dpp/package.nix:42]: for darwin this path expects a .dylib for x86_64 a .so
should I remove it?
I also tried it in darwin but I think more testing is required ( worked fine in my m1 mac )

File: pkgs/by-name/dp/dpp/package.nix (hunk 14 above)
Comment (Xbz-24) [reply pkgs/by-name/dp/dpp/package.nix]: removed now

File: pkgs/by-name/dp/dpp/package.nix (hunk 12 above)
Comment (Xbz-24) [reply pkgs/by-name/dp/dpp/package.nix]: done

File: pkgs/development/libraries/dpp/default.nix
Comment (Xbz-24) [reply pkgs/development/libraries/dpp/default.nix:1]: I thought I did, but the by-name test proved me wrong, what else is needed here?

File: pkgs/by-name/dp/dpp/package.nix (hunk 13 above)
Comment (dotlambda) [reply pkgs/by-name/dp/dpp/package.nix:41]: CMake should be able to figure that out automatically.

**Review by dotlambda — CHANGES_REQUESTED** [review]
Please squash and turn into two commits

* `maintainers: add xbz`
* `dpp: init at 10.0.29`

File: pkgs/top-level/all-packages.nix [hunk 15]
```diff
@@ -624,6 +624,8 @@ with pkgs;
 
   domine = callPackage ../tools/misc/domine { };
 
+  dpp = callPackage ../by-name/dp/dpp/package.nix { };
```
Comment (dotlambda) [inline pkgs/top-level/all-packages.nix:627]:
```suggestion
```

File: pkgs/by-name/dp/dpp/package.nix [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+  src = fetchFromGitHub {
+    owner = "brainboxdotcc";
+    repo = "DPP";
+    rev = "v${version}";
+    hash = "sha256-BJMg3MLSfb9x/2lPHITeI3SWwW1OZVUUMVltTWUcw9I=";
+  };
+
+  nativeBuildInputs =  [
+    cmake
+    libopus
```
Comment (dotlambda) [inline pkgs/by-name/dp/dpp/package.nix:26]:
```suggestion
```

**Review by dotlambda — COMMENTED** [review]
Please squash and turn into two commits

* `maintainers: add xbz`
* `dpp: init at 10.0.29`

**Comment by Xbz-24** [comment]
> Please squash and turn into two commits
> 
> * `maintainers: add xbz`
> * `dpp: init at 10.0.29`

I think I messed up the squash now the label python got added

File: pkgs/by-name/dp/dpp/package.nix (hunk 16 above)
Comment (Xbz-24) [reply pkgs/by-name/dp/dpp/package.nix]: removed

File: pkgs/top-level/all-packages.nix (hunk 15 above)
Comment (Xbz-24) [reply pkgs/top-level/all-packages.nix]: removed now

**Comment by dotlambda** [comment]
> I think I messed up the squash now the label python got added

Do `git rebase -i HEAD~6` and remove the superfluous commits.

**Comment by Xbz-24** [comment]
> > I think I messed up the squash now the label python got added
> 
> Do `git rebase -i HEAD~6` and remove the superfluous commits. Also, adding the maintainer should come before adding the package.

I did but I forgot about the order, do I rebase and change the order and push once again?

**Comment by dotlambda** [comment]
> do I rebase and change the order and push once again?

Yes, but there are still superfluous commits.

**Review by dotlambda — CHANGES_REQUESTED** [review]
Make sure the changes are actually made by the respective commit, not all by one commit.

**Comment by dotlambda** [comment]
Please stop pushing stuff that doesn't address the review comments.

**Comment by Xbz-24** [comment]
> Please stop pushing stuff that doesn't address the review comments.

sorry about the mess I made. Please let me know if it needs further changes

File: pkgs/by-name/dp/dpp/package.nix [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+  ];
+
+  meta = {
+    description = "Discord C++ Library";
+    homepage = "https://github.com/brainboxdotcc/DPP";
+    license = lib.licenses.asl20;
+    maintainers = with lib.maintainers; [ xbz ];
+  };
+}
```
Comment (arthsmn) [inline pkgs/by-name/dp/dpp/package.nix:42]:
```suggestion
stdenv.mkDerivation (finalAttrs: {
  pname = "dpp";
  version = "10.0.29";

  src = fetchFromGitHub {
    owner = "brainboxdotcc";
    repo = "DPP";
    rev = "v${finalAttrs.version}";
    hash = "sha256-BJMg3MLSfb9x/2lPHITeI3SWwW1OZVUUMVltTWUcw9I=";
  };

  nativeBuildInputs = [
    cmake
    pkg-config
  ] ++ lib.optional (stdenv.hostPlatform.isLinux) autoPatchelfHook;

  buildInputs = [
    openssl
    zlib
    libsodium
    libopus
  ];

  meta = {
    description = "Discord C++ Library";
    homepage = "https://github.com/brainboxdotcc/DPP";
    license = lib.licenses.asl20;
    maintainers = with lib.maintainers; [ xbz ];
  };
})
```

Comment (Xbz-24) [reply pkgs/by-name/dp/dpp/package.nix]: done, thank you for your input

Comment (arthsmn) [reply pkgs/by-name/dp/dpp/package.nix]: @Xbz-24 just as a tip, you can resolve the conversations you have already dealt with, this makes the thread cleaner.

**Review by dotlambda — COMMENTED** [review]
I think it makes sense to add `meta.longDescription` where you mention "D++" so that term is greppable.

File: pkgs/by-name/dp/dpp/package.nix [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+    owner = "brainboxdotcc";
+    repo = "DPP";
+    rev = "v${finalAttrs.version}";
+    hash = "sha256-BJMg3MLSfb9x/2lPHITeI3SWwW1OZVUUMVltTWUcw9I=";
+  };
+
+  nativeBuildInputs = [
+    cmake
+    pkg-config
+  ] ++ lib.optional (stdenv.hostPlatform.isLinux) autoPatchelfHook;
```
[the remark this answers is not in the document]
Comment (Xbz-24) [reply pkgs/by-name/dp/dpp/package.nix:27]: autoPatchelfHook is not present in darwin

File: pkgs/by-name/dp/dpp/package.nix [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+    zlib
+    libsodium
+    libopus
+  ];
+
+  meta = {
+    description = "Discord C++ Library";
+    homepage = "https://github.com/brainboxdotcc/DPP";
+    license = lib.licenses.asl20;
+    maintainers = with lib.maintainers; [ xbz ];
```
[the remark this answers is not in the document]
Comment (Xbz-24) [reply pkgs/by-name/dp/dpp/package.nix:40]: if I will be recoming optional auto patch elf hook then I will need only to support x86_64

File: pkgs/by-name/dp/dpp/package.nix [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+
+  buildInputs = [
+    openssl
+    zlib
+    libsodium
+    libopus
+  ];
+
+  meta = {
+    description = "Discord C++ Library";
```
[the remark this answers is not in the document]
Comment (Xbz-24) [reply pkgs/by-name/dp/dpp/package.nix:37]: done

[the remark this answers is not in the document]
Comment (Xbz-24) [reply pkgs/by-name/dp/dpp/package.nix:37]: done

File: pkgs/by-name/dp/dpp/package.nix [hunk 21]
```diff
@@ -0,0 +1,42 @@
+{ stdenv
+, fetchFromGitHub
+, cmake
+, libopus
+, openssl
+, zlib
+, libsodium
+, gcc
```
[the remark this answers is not in the document]
Comment (Xbz-24) [reply pkgs/by-name/dp/dpp/package.nix]: removed, thanks

**Comment by Xbz-24** [comment]
> I think it makes sense to add `meta.longDescription` where you mention "D++" so that term is greppable.

changes where made addressing this in the last commit, please let me know what else needs to be modified

File: pkgs/by-name/dp/dpp/package.nix [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+      D++ (DPP) is a lightweight and simple library for Discord written in modern C++.
+      It is designed to cover as much of the API specification as possible and to have
+      an incredibly small memory footprint, even when caching large amounts of data.
+      This package contains version ${finalAttrs.version} of the DPP library.
+    '';
+    homepage = "https://github.com/brainboxdotcc/DPP";
+    changelog = "https://github.com/brainboxdotcc/DPP/releases/tag/${finalAttrs.src.rev}";
+    license = lib.licenses.asl20;
+    maintainers = with lib.maintainers; [ xbz ];
+    platforms = [ "x86_64-linux" ];
```
Comment (dotlambda) [inline pkgs/by-name/dp/dpp/package.nix:47]:
```suggestion
    platforms = lib.platforms.linux;
```
should be fine

Comment (Xbz-24) [reply pkgs/by-name/dp/dpp/package.nix]: done

**Review by dotlambda — APPROVED** [review]
I rebased for you.
